### PR TITLE
[MIR] Refine representation and translation of calls

### LINKED ISSUE #29573: Mir: remove Panic terminator and instead just call suitable lang item
Panic in MIR is unimplemented. It should call into the panic lang item.

### LINKED ISSUE #29767: Mir: Refactor calls to permit calls that cannot panic and to better accommodate calls that always panic
Today, calls are always basic block terminators. This is unnecessary for calls that cannot panic. We might consider adding a call rvalue. The plus side is that it permits fewer basic blocks and in particular permits avoiding the unwinding code. The minus side is that it means more than one way for a call to happen.

Another thing is that calls today always have two successors. But for fns of type `-> !`, they never return normally, so they should only have a panic successor.

It might be better to leave calls as always being terminators, but just have the number of successors be either 1 or 2. If the number of successors is 1, then the fn is either diverging (`-> !`), in which case this is a panic successor, or else it means the fn is known not to panic somehow. 

cc @Aatch you were mentioning this

### LINKED ISSUE #30480: Mir: translation of Lvalue::ReturnPointer ICEs for functions without return type
```
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:330
stack backtrace:
   1:     0x7f09530f98de - sys::backtrace::tracing::imp::write::h7557e2ce169dfa4119t
   2:     0x7f09531009c5 - panicking::log_panic::closure.41813
   3:     0x7f0953100465 - panicking::log_panic::h56b78895bafbce42lhy
   4:     0x7f09530c78e3 - sys_common::unwind::begin_unwind_inner::h6d6c187a6f5c73b5X0s
   5:     0x7f09530c8267 - sys_common::unwind::begin_unwind_fmt::h4af1494f5de9b7cb3Zs
   6:     0x7f09530f7771 - rust_begin_unwind
   7:     0x7f095311283f - panicking::panic_fmt::h307c1afa6e9b377breJ
   8:     0x7f095310cd48 - panicking::panic::h00907b475580f497YcJ
   9:     0x7f0951b7bb87 - trans::mir::lvalue::MirContext<'bcx, 'tcx>::trans_lvalue::ha96b32c5da76fc05ruO
  10:     0x7f0951a90073 - trans::mir::trans_mir::h96dd1e8056a72b5clNN
  11:     0x7f0951a8a998 - trans::base::trans_closure::h9b514d50ac62ad26vdj
  12:     0x7f0951a92cbc - trans::base::trans_fn::h1126368745048aa2Enj
  13:     0x7f0951a97285 - trans::base::trans_item::ha15e0b1de8fab2e8rNj
  14:     0x7f0951aa38c0 - trans::base::trans_crate::h8f18fa6f69635785zzk
  15:     0x7f095362fe7a - driver::phase_4_translate_to_llvm::hc6d3ed6089be6c3e7Ua
  16:     0x7f0953629c17 - driver::phase_3_run_analysis_passes::closure.23288
  17:     0x7f0953607965 - middle::ty::context::ctxt<'tcx>::create_and_enter::h14213449263107283385
  18:     0x7f09536030c9 - driver::phase_3_run_analysis_passes::h13029203733993090974
  19:     0x7f09535e6c8d - driver::compile_input::h3952c7f29b895d23jca
  20:     0x7f0953748ecb - run_compiler::h1bb4927bc0bc5681Vyc
  21:     0x7f09537460f8 - sys_common::unwind::try::try_fn::h16299814252884817712
  22:     0x7f09530f75d8 - __rust_try
  23:     0x7f09530ee66b - sys_common::unwind::try::inner_try::h5a09ecc8aa81bdabvXs
  24:     0x7f0953746420 - boxed::F.FnBox<A>::call_box::h18362388286111528028
  25:     0x7f09530fec53 - sys::thread::Thread::new::thread_start::h66f81bdcccd0ead1yjx
  26:     0x7f094cdd74a3 - start_thread
  27:     0x7f0952d8c13c - clone
  28:                0x0 - <unknown>
```

This happens because functions like 

```
fn nil(){}
```

have no return type, but `rustc_trans::mir::lvalue::trans_lvalue` unwraps the option when translating the ReturnPointer Lvalue.

### PR DESCRIPTION
**Pull request by nagisa**
r? @nikomatsakis 

This is a pretty big PR conflating changes to a few different block terminators (Call, DivergingCall, Panic, Resume, Diverge), because they are somewhat closely related.

Each commit has a pretty good description on what is being changed in each commit. The end result is greatly simplified CFG and translation for calls (no success branch if the function is diverging, no cleanup branch if there’s nothing to cleanup etc).

Fixes https://github.com/rust-lang/rust/issues/30480
Fixes https://github.com/rust-lang/rust/issues/29767
Partialy solves https://github.com/rust-lang/rust/issues/29575
Fixes https://github.com/rust-lang/rust/issues/29573

### CODE REVIEW

File: src/librustc_trans/trans/mir/block.rs [hunk 1]
```diff
@@ -143,27 +128,100 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
...
+                                      landingpad.llbb,
+                                      Some(attrs),
+                                      debugloc);
                     }
                 }
+            }
+        }
+    }
+
+    fn make_landing_pad(&mut self, cleanup: mir::BasicBlock) -> Block<'bcx, 'tcx> {
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/block.rs]: Note to self: Should not ignore the `-C no-landing-pads` option.

File: src/librustc_trans/trans/mir/block.rs [hunk 2]
```diff
@@ -143,27 +128,100 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
...
+        build::Br(bcx, self.llblock(cleanup), DebugLoc::None);
+        bcx
+    }
 
-                build::Br(bcx, self.llblock(targets.0), DebugLoc::None)
+    fn unreachable_block(&mut self) -> Block<'bcx, 'tcx> {
+        match self.unreachable_block {
+            Some(b) => b,
+            None => {
+                let bl = self.bcx(mir::START_BLOCK).fcx.new_block(false, "unreachable", None);
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/block.rs]: Note to self: maybe `self` should contain a fcx too?

File: src/librustc_mir/build/scope.rs [hunk 3]
```diff
@@ -267,29 +288,94 @@ impl<'a,'tcx> Builder<'a,'tcx> {
...
-    if len == 0 {
-        return DIVERGE_BLOCK;
+    pub fn panic_bound_check(&mut self,
+                             block: BasicBlock,
+                             index: Operand<'tcx>,
+                             len: Operand<'tcx>,
+                             span: Span) {
+        let cleanup = self.diverge_cleanup();
+        let func = self.lang_function(lang_items::PanicBoundsCheckFnLangItem);
+        let str_ty = self.hir.tcx().mk_static_str();
```
Comment (nagisa) [inline src/librustc_mir/build/scope.rs]: I really dislike constructing types manually here. Is there another way?

File: src/librustc_mir/build/scope.rs [hunk 4]
```diff
@@ -267,29 +288,94 @@ impl<'a,'tcx> Builder<'a,'tcx> {
...
         self.scopes.first().map(|scope| scope.extent).unwrap()
     }
-}
 
-fn diverge_cleanup_helper<'tcx>(cfg: &mut CFG<'tcx>, scopes: &mut [Scope<'tcx>]) -> BasicBlock {
-    let len = scopes.len();
-
-    if len == 0 {
-        return DIVERGE_BLOCK;
+    pub fn panic_bound_check(&mut self,
```
Comment (nagisa) [inline src/librustc_mir/build/scope.rs]: Cannot be tested for correctness at the moment because https://github.com/rust-lang/rust/issues/30474

File: src/librustc_mir/build/matches/mod.rs [hunk 5]
```diff
@@ -92,7 +92,7 @@ impl<'a,'tcx> Builder<'a,'tcx> {
         // not entirely precise
         if !otherwise.is_empty() {
             let join_block = self.join_otherwise_blocks(otherwise);
-            self.panic(join_block);
+            self.panic(join_block, "something about matches algorithm not being precise", span);
```
Comment (nagisa) [inline src/librustc_mir/build/matches/mod.rs]: This looks, uh, ugly? Incorrect?

**Comment by nagisa** [comment]
All in all, I think it is in a good enough spot now to begin reviewing it @nikomatsakis.

File: src/librustc_trans/trans/mir/block.rs [hunk 6]
```diff
@@ -105,65 +96,140 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
...
+                        let landingpad = self.make_landing_pad(cleanup);
+                        build::Invoke(bcx,
+                                      callee.immediate(),
+                                      &llargs[..],
+                                      self.llblock(targets.0),
+                                      landingpad.llbb,
+                                      Some(attrs),
+                                      debugloc);
+                        if must_copy_dest {
+                            // FIXME: What do we do here?
```
Comment (luqmana) [inline src/librustc_trans/trans/mir/block.rs]: AFAIK should just do the same as below: store_ty the result from Invoke to the destination.

Comment (nagisa) [inline src/librustc_trans/trans/mir/block.rs]: The problem here (is much harder than it looks at the first look!) is that `invoke` is a basic block terminator and thus we cannot add anything else to the current block. We would have to add, or rather, *prepend* a store instruction to a successful branch, but… that’s tricky, because:

1. That’s hacky (duh);
2. I can’t seem to see anyway to do it;
3. It would make some already-generated blocks invalid (e.g. we would have to assert that success branch has higher basic-block number).

Basically, to me it seems like we should change the MIR some more and figure out if destination must be copied (or rather, avoid copying destination altogether) during the MIR build phase.

**Comment by nikomatsakis** [comment]
As I mentioned on IRC, I'm not sure if I think that splitting out "diverging call" into its own terminator is a good idea. My concerns are:

1. Duplication: it's harder to know that you're getting all calls.
2. If you have `foo = diverging_fn()`, we need to still consider that a "write" to `foo` in a logical sense (i.e., for the purposes of borrowck). To do that under the approach of this PR, we'll need some marker that `foo` has been initialized, since `DivergingCall` no longer has a destination.

Basically, I'm not sure if the code for handling calls vs diverging calls will wind up being so very different in other passes. But it's a bit hard to know just now. A compromise might just be to introduce an enum into the call that is either `NormalCall(BasicBlock, BasicBlock)` or `DivergingCall(BasicBlock)`.

In any case, let me read further before reaching final judgement.

**Comment by nagisa** [comment]
> As I mentioned on IRC, I'm not sure if I think that splitting out "diverging call" into its own terminator is a good idea. My concerns are:

The last commit contains conversion to inner-enum, like you proposed on IRC. Feel free to compare.

**Comment by nagisa** [comment]
Comments on https://github.com/nagisa/rust/commit/0c7d57d24456d6b306a2a71dee7d3393187dfccf are yet to be fixed.

**Comment by nikomatsakis** [comment]
In all those cases where I gave nits about exhaustive matches, we were just trying to handle converging cases -- might be that a helper function for extracting the converging cases would be good. OTOH, I guess this is why you split those cases into two to start with.

**Comment by nikomatsakis** [comment]
OK, so, as you say, this is a big PR, and I feel like I could go on pointing out nits forever. But I'd rather we land it and just hack and try to clean things up. Some of these commits get pretty involved, but that may just be unavoidable if we want to generate less naive IR. Calls are just complicated. :) Anyway, so I'm leaning now towards landing this PR and then trying to keep improving things. One thing which would be nice are some more test cases for calls that generate `invoke`, and in particular to exercise some of those edge cases paths (like `loop { c = call(); }`).

**Comment by nagisa** [comment]
@nikomatsakis rebased and comments addressed (I hope I didn’t miss any). I added a few more tests as well.

**Comment by nikomatsakis** [comment]
r=me modulo final nits (in particular, the test with graphviz attribute should really be fixed)

File: src/librustc/mir/repr.rs [hunk 7]
```diff
@@ -243,40 +235,88 @@ pub enum Terminator<'tcx> {
...
+            CallKind::DivergingCleanup(ref mut b) |
+            CallKind::Converging { target: ref mut b, .. } => slice::mut_ref_slice(b),
+            CallKind::ConvergingCleanup { ref mut targets, .. } => targets.as_mut_slice(),
+        }
+    }
+
+    pub fn destination(&self) -> Option<Lvalue<'tcx>> {
+        match *self {
+            CallKind::Converging { ref destination, .. } |
+            CallKind::ConvergingCleanup { ref destination, .. } => Some(destination.clone()),
```
Comment (tsion) [inline src/librustc/mir/repr.rs]: Why clone here instead of returning `Option<&Lvalue<'tcx>>` and letting the caller decide if it needs to clone?

Comment (nagisa) [inline src/librustc/mir/repr.rs]: There’s two reasons, mostly, I had in mind:

1. Cloning `Lvalue` is very cheap most of the time (i.e. when `Lvalue` is not a `Projection`);
2. There’s users who want `&mut lvalue` and there’s users who want `&lvalue`. Returning a value allows to make either one easier when pattern matching (`ref` vs `ref mut`). 

On the hindsight, I think this might actually be a bug to return destination by value in some cases which I didn’t think about before, thus I’ll submit a patch in a moment making these return a mutable reference instead.
